code/modules/mob/living/simple_animal/friendly/trader.dm

```diff
@@ -94,154 +181,285 @@
+ * Arguments:
+ * * user - (Mob REF) The mob trying to sell something
+ */
+/mob/living/simple_animal/hostile/retaliate/trader/proc/try_sell(mob/user)
+	var/obj/item/activehanditem = user.get_active_held_item()
+	var/obj/item/inactivehanditem = user.get_inactive_held_item()
+	if(!sell_item(user, activehanditem) && !sell_item(user, inactivehanditem))
+		say(return_trader_phrase(ITEM_REJECTED_PHRASE))
 
 /**
  * Checks if an item is in the list of wanted items and if it is after a Yes/No radial returns generate_cash with the value of the item for the NPC
  * Arguments:
- * * user - The mob trying to sell something
- * * selling - The item being sold
+ * * user - (Mob REF) The mob trying to sell something
+ * * selling - (Item REF) The item being sold
  */
-/mob/living/simple_animal/hostile/retaliate/trader/proc/sell_item(mob/user, selling)
-	var/obj/item/sellitem = selling
+/mob/living/simple_animal/hostile/retaliate/trader/proc/sell_item(mob/user, obj/item/selling)
 	var/cost
-	if(!sellitem)
+	if(!selling)
+		return FALSE
+	var/list/product_info
+	//Keep track of the typepath; rather mundane but it's required for correctly modifying the wanted_items
+	//should a product be sellable because even if it doesn't have a entry because it's a child of a parent that is present on the list
+	var/typepath_for_product_info
+	if(selling.type in wanted_items)
+		product_info = wanted_items[selling.type]
+		typepath_for_product_info = selling.type
+	else //Assume wanted_items is setup in the correct way; read wanted_items documentation for more info
+		for(var/typepath in wanted_items)
+			if(istype(selling, typepath))
+				product_info = wanted_items[typepath]
+				typepath_for_product_info = typepath
+				break
+
+	if(!product_info) //Nothing interesting to sell
+		return FALSE
+	if(product_info[TRADER_PRODUCT_INFO_QUANTITY] <= 0)
+		say(return_trader_phrase(TRADER_HAS_ENOUGH_ITEM_PHRASE))
 		return FALSE
-	var/datum/checked_type = sellitem.type
-	do
-		if(checked_type in wanted_items)
-			cost = wanted_items[checked_type]
-			break
-		checked_type = checked_type.parent_type
-	while(checked_type != /obj)
-	if(!cost)
+	cost = apply_sell_price_mods(selling, product_info[TRADER_PRODUCT_INFO_PRICE])
+	if(cost <= 0)
+		say(return_trader_phrase(ITEM_IS_WORTHLESS_PHRASE))
 		return FALSE
-	say(interestedphrase)
-	to_chat(user, span_notice("You will receive [cost] credits for each one of [sellitem]."))
+	say(return_trader_phrase(INTERESTED_PHRASE))
+	to_chat(user, span_notice("You will receive [cost] [currency_name] for each one of [selling]."))
 	var/list/npc_options = list(
 		"Yes" = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_yes"),
 		"No" = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_no")
 	)
 	var/npc_result = show_radial_menu(user, src, npc_options, custom_check = CALLBACK(src, .proc/check_menu, user), require_near = TRUE, tooltips = TRUE)
 	face_atom(user)
```

Comment:
```suggestion
	face_atom(user)
	var/npc_result = show_radial_menu(user, src, npc_options, custom_check = CALLBACK(src, .proc/check_menu, user), require_near = TRUE, tooltips = TRUE)
```

Reply:
Pending on this being used as a fix for a runtime